module/zfs/zfs_vnops.c

```diff
@@ -202,6 +210,77 @@ zfs_access(znode_t *zp, int mode, int flag, cred_t *cr)
 	return (error);
 }
 
+/*
+ * Determine if Direct I/O has been requested (either via the O_DIRECT flag or
+ * the "direct" dataset property). When inherited by the property only apply
+ * the O_DIRECT flag to correctly aligned IO requests. The rational for this
+ * is it allows the property to be safely set on a dataset without forcing
+ * all of the applications to be aware of the alignment restrictions. When
+ * O_DIRECT is explicitly requested by an application return EINVAL if the
+ * request is unaligned.  In all cases, if the range for this request has
+ * been mmap'ed then we will perform buffered I/O to keep the mapped region
+ * synhronized with the ARC.
+ *
+ * It is possible that a file's pages could be mmap'ed after it is checked
+ * here. If so, that is handled coorarding in zfs_write(). See comments in the
+ * following area for how this is handled:
+ * zfs_write() -> update_pages()
+ */
+static int
+zfs_setup_direct(struct znode *zp, zfs_uio_t *uio, zfs_uio_rw_t rw,
+    int *ioflagp)
+{
+	zfsvfs_t *zfsvfs = ZTOZSB(zp);
+	objset_t *os = zfsvfs->z_os;
+	int ioflag = *ioflagp;
+	int error = 0;
+
+	if (!zfs_dio_enabled || os->os_direct == ZFS_DIRECT_DISABLED ||
+	    zn_has_cached_data(zp, zfs_uio_offset(uio),
+	    zfs_uio_offset(uio) + zfs_uio_resid(uio) - 1)) {
+		/*
+		 * Direct I/O is disabled or the region is mmap'ed. In either
+		 * case the I/O request will just directed through the ARC.
+		 */
+		ioflag &= ~O_DIRECT;
+		goto out;
+	} else if (os->os_direct == ZFS_DIRECT_ALWAYS &&
+	    zfs_uio_page_aligned(uio) &&
+	    zfs_uio_aligned(uio, PAGE_SIZE)) {
+		if ((rw == UIO_WRITE && zfs_uio_resid(uio) >= zp->z_blksz) ||
+		    (rw == UIO_READ)) {
+			ioflag |= O_DIRECT;
+		}
+	} else if (os->os_direct == ZFS_DIRECT_ALWAYS && (ioflag & O_DIRECT)) {
+		/*
+		 * Direct I/O was requested through the direct=always, but it
+		 * is not properly PAGE_SIZE aligned. The request will be
+		 * directed through the ARC.
+		 */
+		ioflag &= ~O_DIRECT;
```

Comment:
So, this is an interesting case. With the `direct` property setting being set to `always`, it is a best attempt at doing acting as though `O_DIRECT` was pass on every `open` call. However, the main thing with the `always` property setting is knowing that alignment restrictions are relaxed. By relaxed, that means that unaligned `PAGE_SIZE` I/O just happily gets passed off to the ARC. This was the entire idea behind `always`. `EINVAL` could never be returned for misalignment. If the users want that kind of strict enforcement of alignment checks, we happily accommodate that. That is the `direct` property being set to `standard`. In my opinion, the semantics of the dataset property supersede the fact that `O_DIRECT` is passed as flag. Otherwise, there would be no reason to even have the `always` property value. It is a best attempt, but assurance that it will not fail under misalignment.